numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -590,7 +590,12 @@ typedef struct {
         PyDataType_FLAGCHK(dtype, NPY_ITEM_REFCOUNT)
 
 typedef struct _PyArray_Descr {
+#ifdef USE_DTYPE_AS_PYOBJECT
         PyObject_HEAD
+#else
+        /* may need to be PyHeapTypeObject for new tp_as_* functions? */
+        PyTypeObject descrtype;
```

Comment:
Yes, this needs to be PyHeapTypeObject.  The Cython team explored this and ran into the problem before until they made meta-types inherit from PyHeapTypeObject.  See, for example, https://mail.python.org/pipermail/cython-devel/2013-September/003813.html
```suggestion
        PyHeapTypeObject descrtype;
```

Reply:
I'm not sure this applies to us. From that thread:

> However, when MyClass is created, an instance of struct __pyx_obj_4meta_MetaClass (cc. 200 + 4 bytes) is dynamically allocated by the Python memory management machinery. The machinery then tries to initialize the allocated memory. 

In our case, the instance of the dtype is `malloc`'d and initialized by `arraydescr_new`, so we're in complete control.

As far as I can tell, the main purpose of `PyHeapTypeObject` is to convert builtin magic methods into slots, something we probably don't care about too much for dtypes.